modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+      rescue StandardError => e
+        fail_with(Failure::BadConfig, "Could not compile the filter to find the ESC1 objects! Error was #{e}")
+      end
+
+      returned_entries = ldap.search(base: full_base_dn, filter: filter, attributes: attributes)
+      query_result = ldap.as_json['result']['ldap_result']
+
+      result_code, result_message = check_query_result_code(query_result, filter)
+      case result_code
+      when -1
+        fail_with(Failure::BadConfig, result_message)
+      when 0
+        vprint_good(result_message)
+      when 1
+        fail_with(Failure::NoAccess, result_message)
+      when 2
+        fail_with(Failure::UnexpectedReply, result_message)
+      end
+
+      if returned_entries.blank?
+        print_error("No results found for #{filter}.")
+
+        nil
+      else
+
+        returned_entries
+      end
+    end
+  rescue Rex::ConnectionTimeout
+    fail_with(Failure::Unreachable, "Couldn't reach #{datastore['RHOST']}!")
+  rescue Net::LDAP::Error => e
+    fail_with(Failure::UnexpectedReply, "Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+  end
+
+  def convert_sids_to_human_readable_name(sids_array)
+    output = []
+    for sid in sids_array
+      raw_filter = "(objectSID=#{sid})"
+      attributes = ['dn', 'description']
+      sid_entry = query_ldap_server(raw_filter, attributes)
+      fail_with(Failure::UnexpectedReply, 'Could not find any matching SID!') if sid_entry.blank?
+      if sid_entry[0][:description]
+        output << [sid, sid_entry[0][:dn][0], sid_entry[0][:description][0]]
+      else
+        output << [sid, sid_entry[0][:dn][0], nil]
+      end
+    end
+
+    result = ''
+    output.each do |sid_string, dn, description|
+      if description
+        result += "#{sid_string} aka #{dn} (#{description}) | "
+      else
+        result += "#{sid_string} aka #{dn} | "
+      end
+    end
+
+    result.gsub(%r{ \| $}, '') # Strip out the last trailing " \ " from the output.
+  end
+
+  def find_esc1_vuln_cert_templates
+    esc1_raw_filter = '(&'\
+      '(objectclass=pkicertificatetemplate)'\
+      '(!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))'\
+      '(|(mspki-ra-signature=0)(!(mspki-ra-signature=*)))'\
+      '(|'\
+        '(pkiextendedkeyusage=1.3.6.1.4.1.311.20.2.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.5.7.3.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.2.3.4)'\
+        '(pkiextendedkeyusage=2.5.29.37.0)'\
+        '(!(pkiextendedkeyusage=*))'\
+      ')'\
+      '(mspki-certificate-name-flag:1.2.840.113556.1.4.804:=1)'\
+    ')'
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc1_entries = query_ldap_server(esc1_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC1 vulnerable certificate templates!') if esc1_entries.blank?
```

Comment:
Don't you want it to continue on to check for ESC2 and ESC3? Doesn't make sense to fail here entirely.

Reply:
Removed early termination code in https://github.com/rapid7/metasploit-framework/pull/17122/commits/7a2ef23755565bf65de53f21440c00cb4c30325f